tensorflow/stream_executor/cuda/cuda_helpers.h

```diff
@@ -22,4 +22,14 @@ limitations under the License.
 
 #include "tensorflow/stream_executor/gpu/gpu_helpers.h"
 
+namespace stream_executor {
+namespace cuda {
+
+// A helper function to decide whether to enable deterministic cuDNN
+// functionality.
+bool RequireCuDNNDeterminism();
```

Comment:
It seems weird for TF and XLA to ask for flags from StreamExecutor, as StreamExecutor is just a library, and it should impose constraints on users about how to use this library.

Ideally, the source of truth of "users require determinism" should be a flag in TF, and then it get passed down to XLA. Both TF and XLA pass down the same boolean to StreamExecutor. But this ideal solution creates unnecessary amount of work.

How about this:
* Create a TF flag TF_DETERMINISTIC_AUTOTUNING in `tensorflow/core/kernel/gpu_utils.h`, which is only used by TF kernels.
* Create a similar XLA flag to DebugOptions in `tensorflow/compiler/xla/xla.proto`, which is used by XLA.
* Keep the current StreamExecutor flags as it is privately in cuda_dnn.cc.

When the end user wants determinism, they have to compose all necessary flags. This way, at least the layering is less coupled.

Reply:
### Adding and Modifying Switches

I strongly prefer to not add additional determinism switches and/or change the functionality of the existing switches. `TF_CUDNN_DETERMINISTIC` makes cuDNN-based ops function deterministically, and `TF_DETERMINISTIC_OPS` aspires to making all ops, including those based on cuDNN, operate deterministically. Since a model either operates deterministically or not (there being no useful gray area), `TF_DETERMINISTIC_OPS` supersedes, and ultimately makes fully redundant, `TF_CUDNN_DETERMINISTIC`.

At the moment `TF_DETERMINISTIC_OPS` is being listened to, like a global variable, in both Python code and C++ code. I'm currently working on a deterministic mode for another op, which will also be enabled by `TF_DETERMINISTIC_OPS`. In time, `TF_DETERMINISTIC_OPS` will be referenced in more and more places in the code-base. I envision the functionality of `TF_DETERMINISTIC_OPS` eventually being replaced by a fully plumbed-in config variable.

I have been communicating the status and recipes for TensorFlow determinism [here](https://github.com/NVIDIA/tensorflow-determinism), and I don't want to make that information unnecessarily complicated for the user. Note that I intend to document TensorFlow determinism at a high level in the official TensorFlow documentation soon.

### Ideal Solution

I'm not sure what you mean by "a flag in TF." Perhaps you're referring to adding something to `tf.config`? Or are you just referring to what is represented by `RequireCuDNNDeterminism()` (the logical OR of `TF_CUDNN_DETERMINISTIC` with `TF_DETERMINISTIC_OPS`)?

Following your outline above, I'm thinking that a practical way forward might be to move `RequireCuDNNDeterminism()` from `tensorflow/stream_executor/cuda/cuda_helpers.h/.cc` to `tensorflow/core/kernels/gpu_utils.h/.cc` and use it where needed in `tensorflow/core/kernels/*` including passing it down in the interface to the GetConvolve\*Algorithms() methods in the stream executor.

I currently don't know how to pass this flag down to XLA, but I could work on trying to figure it out. Then XLA could also use it and pass it to the GetConvolve\*Algorithms() methods in the stream executor.

### Compromise

Would it be possible and acceptable for me to replicate the `RequireCuDNNDeterminism()` code (which looks at both `TF_CUDNN_DETERMINISM` and `TF_DETERMINISTIC_OPS`) in `tensorflow/compiler/xla/service/gpu/gpu_conv_algorithm_picker.cc`? There, the setting can be both used and passed down to GetConvolve\*Algorithms() methods in the stream executor.

Reply:
I just pushed a commit that removes the dependency of both `core/kernels:gpu_utils` and `xla/service/gpu:gpu_conv_algorithm_picker` on `stream_executor`. This has been done simply by replicating the code that listens for, and caches, `TF_CUDNN_DETERMINISTIC || TF_DETERMINISTIC_OPS` in both `core/kernels/gpu_utils.cc` and `xla/service/gpu/gpu_conv_algorithm_picker.cc`

I considered passing a `require_determinism` flag through the parameter lists of the `CudnnSupport::GetConvolve\*Algorithms()` methods (defined in `stream_executor/cuda/cuda_dnn.cc`), rather than generating the flag locally within `cuda_dnn.cc`, but realized that the flag was also needed for selecting the deterministic operation of cuDNN max-pooling, which is also done in `cuda_dnn.cc`. There was no easy way to propagate the flag in that case.

So currently, the flag is generated independently, and equivalently, in three different places. To prevent code repetition, I would prefer for it to be in a library that can be included by stream_executor, core, and XLA. I don't know what would be the right library for that, however. I originally incorrectly chose `stream_executor/cuda:cuda_helpers` because of my lack of understanding of the architecture of TensorFlow.

To re-summarize my intentions:

1. I want to keep the API for determinism as simple as possible for users, which means not changing the meaning or functionality of `TF_CUDNN_DETERMINISTIC` and `TF_DETERMINISTIC_OPS`, and also extending the promise that they represent. This pull request represents a bug fix that requires those environment variables to directly control code in XLA and `core/kernels`.

2. For this development work, I think it makes sense to be able to make the changes where necessary to get the deterministic functionality as quickly as possible while maintaining a simple and consistent API. Going forward, this is mostly going to appear as different parts of the codebase listening directly to `TF_DETERMINISTIC_OPS`. Ultimately, either myself, or someone else, can go though and properly plumb the switch into all the places that have been highlighted by the use of `TF_DETERMINISTIC_OPS`, possibly from something like `tf.config.deterministic_ops`.

Reply:
Assuming that it would be preferable to define `RequireCuDNNDeterminism()` once, in one place, I've been looking for an appropriate place to do that. I'm wondering if `tensorflow/core/common_runtime/gpu` would make sense. I'm thinking that it could go into a new "module" (.h + .cc) called `gpu_determinism`. Would it be okay for stream_executor, core, and XLA to import and use this?

Reply:
I think I've found another (even more?) appropriate place to define `RequireCudnnDeterminism()`: `tensorflow/core/util:use_cudnn` ([use_cudnn.h](https://github.com/tensorflow/tensorflow/blob/v2.0.0/tensorflow/core/util/use_cudnn.h) / [use_cudnn.cc](https://github.com/tensorflow/tensorflow/blob/v2.0.0/tensorflow/core/util/use_cudnn.cc)).

Reply:
Hi @duncanriach ,

Thanks for the detailed survey and context of tensorflow-determinism.

> I have been communicating the status and recipes for TensorFlow determinism here, and I don't want to make that information unnecessarily complicated for the user. 

Agreed. I think we can achieve both "let users control determinism with a single env var" and "use multiple internal global states". I'll describe a revised proposal below.

> I'm not sure what you mean by "a flag in TF." Perhaps you're referring to adding something to tf.config? Or are you just referring to what is represented by RequireCuDNNDeterminism() (the logical OR of TF_CUDNN_DETERMINISTIC with TF_DETERMINISTIC_OPS)?

For the context, I was referring to env vars like TF_CUDNN_DETERMINISTIC as "a flag in TF".

> Ideal Solution
It sounds plausible, with the following complement for a more long term solution, with hopefully reasonable implementation cost:
* Move RequireCuDNNDeterminism() to tensorflow/common_runtime/gpu:gpu_determinism (already described).
* For XLA, create a new DebugOption flag "xla_deterministic_autotuning" [1] and append [2] it from tensorflow/common_runtime/gpu:gpu_determinism, preferably in a global initializer.
* For StreamExecutor, create a `virtual void DnnSupport::SetRequireDeterminism()`, and call it from the global initializer in gpu_determinism. `SetRequireDeterminism()` sets a bool member in CudnnSupport that turns on and off determinism. DnnSupport is a singleton and can be accessed from `StreamExecutor::AsDnn()`.

[1] https://github.com/tensorflow/tensorflow/blob/cf8929973986d1eeac6acc10cb36b17075ca075e/tensorflow/compiler/xla/xla.proto#L25
[2] https://github.com/tensorflow/tensorflow/blob/96de19babd874ea6be7674520e0d8351ca89a749/tensorflow/compiler/xla/debug_options_flags.h#L29

Reply:
Thanks so much for that detailed description, @timshen91. I can see the path to implementing that pretty clearly. I do have a concern with this approach though. A global initializer is required because there is no triggering event to set the internal state. I imagine that a C++ global initializer will run when TensorFlow is imported into Python. Do you know if that is true?

If it's true that the C++ global initializer will run when TensorFlow is imported into Python, then it would be necessary to set `TF_CUDNN_DETERMINISTIC` or `TF_DETERMINISTIC_OPS` before importing TensorFlow. This would be quite a significant change in the API for current users, which would need to be conveyed for versions of TensorFlow beyond a certain point (e.g v2.2) and would also require user code changes (existing code would break). Also, our kernel tests set these environment variables in python-main before starting the tests. I don't know if it's possible, or desirable, to set the environment variables prior to the import of TensorFlow. I don't know how to achieve that in the test files and/or the bazel build config (using currently available functionality).

The way it works currently in the master branch is that the first time the state of one of these environment variables is needed, it gets referenced and cached. The recommended, and most common, usage pattern is to set these environment variables using os.environ after importing TensorFlow and before building or training a graph. I've been trying to replicate this functionality even after the need to reference those environment variables has bled out of stream executor.

Reply:
> Thanks so much for that detailed description, @timshen91. I can see the path to implementing that pretty clearly. I do have a concern with this approach though. A global initializer is required because there is no triggering event to set the internal state. I imagine that a C++ global initializer will run when TensorFlow is imported into Python. Do you know if that is true?

In general, global initializers run when the .so file is dlopen()ed, but I'm not sure when exactly.

> 
> If it's true that the C++ global initializer will run when TensorFlow is imported into Python, then it would be necessary to set `TF_CUDNN_DETERMINISTIC` or `TF_DETERMINISTIC_OPS` before importing TensorFlow. This would be quite a significant change in the API for current users, which would need to be conveyed for versions of TensorFlow beyond a certain point (e.g v2.2) and would also require user code changes (existing code would break).

The global initializer part was more of a straw-proposal, given that I don't know enough about tensorflow-determinism. How about using std::once in RequireDeterminism(), so that the initialization is defered until the first check?

> Also, our kernel tests set these environment variables in python-main before starting the tests. I don't know if it's possible, or desirable, to set the environment variables prior to the import of TensorFlow. I don't know how to achieve that in the test files and/or the bazel build config (using currently available functionality).
> 
> The way it works currently in the master branch is that the first time the state of one of these environment variables is needed, it gets referenced and cached.

This sounds like the std::once behavior I described above.

> The recommended, and most common, usage pattern is to set these environment variables using os.environ after importing TensorFlow and before building or training a graph. I've been trying to replicate this functionality even after the need to reference those environment variables has bled out of stream executor.

Reply:
Are you referring to [`std::call_once`](https://en.cppreference.com/w/cpp/thread/call_once)? I'm not sure that this addresses the potential issue. We currently have a solution for caching the state on the first call and then returning the cached value on subsequent calls. Whether that solution is thread-safe, and whether it needs to be thread-safe, is another issue.

The problem is what triggers the first call to `RequireCudnnDeterminism()`. In the current master-branch, the trigger is operations running inside the stream executor. In the currently-proposed changes, the trigger is operations running inside the TF or XLA kernel code (or inside the stream executor).

To capture the values of the environment variables after TensorFlow is imported but before graph construction and execution is performed, there needs to be a triggering event. If we can't use the actual need for the state as the triggering event, then there needs to be an alternative. If global initializers run when TensorFlow is imported, then they cannot be used.

So, without being able to use the need for the state as the triggering event, the time when the state should be checked, and assuming that using a global initializer will not work (which I'll try to confirm today) then there needs to be some other triggering event in TensorFlow. If the state in both XLA and stream executor is set via `tensorflow/common_runtime/gpu:gpu_determinism` then there would need to be a function in that module which gets called either before any op is invoked or, more sensibly perhaps, before specific ops are invoked. The specific list of ops would be any that uses convolution or max-pooling. Early in the invocation of one of those ops, there would be a call to a function in `tensorflow/common_runtime/gpu:gpu_determinism` to grab the environment variable state and appropriately configure the state in both XLA and stream executor.

Note that all of this is an attempt to model a simple environment variable configuration, which itself is only intended to be a temporary solution until we implement a properly plumbed-in `tf.config.set_deterministic_ops` API in a future version of TensorFlow. When we have `tf.config.set_deterministic_ops` then calling that will be the triggering event.

Actually, what I'm wondering if it makes most sense to now implement `tf.config.set_deterministic_ops` and have that configure XLA, stream executor, and the state returned to the regular TF kernel code via `RequireCudnnDeterminism()`. It would also set the environment variable `TF_DETERMINISTIC_OPS`, which would propagate the state to other places in the codebase that listen to (or will listen to) this environment variable. However, while this would move us forward in terms of implementing a more correct long-term API for this feature, it would not address the issue of the current API getting broken.

My next action is to check when C++ global initializers run relative to the TensorFlow import into Python.

Reply:
Here is the result of the experiment with global initializers. I put the following code into `tensorflow/core/util/use_cudnn.cc` and built a wheel:

```
#include <iostream>

class GlobalInitializer {
public:
  GlobalInitializer() {
    std::cout << "########## RUNNING THE INITIALIZER ##########" << std::endl;
  }
};

GlobalInitializer globalInitializer;
```

After pip installing the wheel, the following behavior was observed in the python interpreter:

```
>>> import tensorflow as tf
########## RUNNING THE INITIALIZER ##########
```

So this confirms that C++ global initializers run when TensorFlow is imported into python. This means that a global initializer cannot be used to emulate the current functionality of `TF_CUDNN_DETERMINISTIC` and `TF_DETERMINISTIC_OPS`.

Reply:
@timshen91, I spent some time looking for locations in the code to trigger the capture of the values of the environment variables `TF_CUDNN_DETERMINISTIC` and `TF_DETERMINISTIC_OPS` and propagate the settings appropriately. I'm not confident that I can reliably identify (or test) all the possible entry points to achieve emulation of the current behavior and maintain the API that is currently documented and that users expect. Even if I could implement and test this, the resulting solution would be incredibly brittle.

If it's not acceptable to fix this bug by having the code directly query the state of the environment variables at the time it's needed (along the lines of the current state of this PR), then I think we should not even attempt to fix this bug at the same time as maintaining backwards compatibility.

The most favorable alternative that I can see is to break compatibility in a future version of TensorFlow in order to fix this bug. For example, in TF version 2.2 we could remove both `TF_CUDNN_DETERMINISTIC` and `TF_DETERMINISTIC_OPS` and instead enable determinism in TensorFlow via `tf.config.experimental.deterministic_ops`. I have no idea how many programs in the wild are currently using the existing environment-variable-based API in the way that has been suggested. Unfortunately, it's not even feasible to display a deprecation warning for them.

I'm increasingly recognizing the shortcomings of configuring via environment variables, especially when they need to be phased out, which has always been the plan. The reason I did it this way in the first place was so that I could focus on achieving deterministic functionality across TensorFlow while not having to worry about the plumbing. Our contacts at Google suggested early in this project that I could/should focus on using environment variables and that they would deal with the plumbing later. Unfortunately, because environment-variable-based initialization is apparently not allowed in the XLA code, we're now in a situation where we cannot fix a bug without breaking backwards compatibility.

If we proceed with the backwards-compatibility-breaking bug fix, then I would like to temporarily close this pull request (possibly for a month or more) so that I can focus on addressing some other urgent sources of non-determinism in TensorFlow. After that, I would hope to be able to return to this PR and carefully push it through to merge.

Reply:
@duncanriach , thanks for looking at these solutions!

I don't have a sense of priority for this PR, so I'm happy to defer the call to you. Depending on the priority, it's fine by me either to hold on this PR (as you seem to suggest) or to query the env var in multiple places (your original commits).

Reply:
If you're willing to take it, and it seems that you are, then I would much prefer to have a bug fix in place immediately, by querying the env var in multiple places (as specified by the current commits).

Please will you clarify what you mean by comments describing the migration path? Do you mean adding comments explaining the intention to migrate to `tf.config.experimental.deterministic_ops` and associated plumbing? If so, I would gladly add that.

Also, are you happy with having the code defined by `RequireCudnnDeterminism()` replicated in three different places in the codebase, or would you prefer, as I would, for it to be defined in one place, such as `tensorflow/core/common_runtime/gpu:gpu_determinism`? Refactoring that would be an easy and quick change to make.

Reply:
> If you're willing to take it, and it seems that you are, then I would much prefer to have a bug fix in place immediately, by querying the env var in multiple places (as specified by the current commits).

Yes, this is what I meant.

> 
> Please will you clarify what you mean by comments describing the migration path? Do you mean adding comments explaining the intention to migrate to `tf.config.experimental.deterministic_ops` and associated plumbing? If so, I would gladly add that.

Yes, plus verbal warning bits like "this is a temporary solution".

> 
> Also, are you happy with having the code defined by `RequireCudnnDeterminism()` replicated in three different places in the codebase, or would you prefer, as I would, for it to be defined in one place, such as `tensorflow/core/common_runtime/gpu:gpu_determinism`? Refactoring that would be an easy and quick change to make.

I prefer to duplicate them in several places all with the comment we talked above, also with "this code is duplicated, and should be in sync with [other files]". I don't expect frequent changes to these duplicates, so keeping them in sync shouldn't be too much work.

Reply:
Got it. Yes, it's very unlikely that the code will ever need to change (until it's removed). I'll work on an incremental commit to address these items. Thank you, Tim.

Reply:
@timshen91, The most recent commit adds a detailed comment to all replicas of the function.